File: webui/src/api/mtc.ts
import { apiJson, apiFetch, apiVoid, ApiError, extractErrorMessage } from './client';

export interface MtcTreeSize {
  tree_size: number;
}

export interface MtcRoot {
  tree_size: number;
  root_hash: string;
}

export interface MtcLandmark {
  sequence_no: number;
  tree_size: number;
  created_at: number;
}

export interface MtcInclusionProof {
  leaf_index: number;
  tree_size: number;
  proof: { hash: string }[];
}

export interface MtcConsistencyProof {
  from_size: number;
  to_size: number;
  from_root: string;
  to_root: string;
}

export interface MtcSubtreeRoot {
  start: number;
  end: number;
  root_hash: string;
}

export interface MtcRevokedRange {
  start: number;
  end: number;
}

function caQuery(caId?: string): string {
  return caId ? `?ca_id=${encodeURIComponent(caId)}` : '';
}

async function fetchText(path: string): Promise<string> {
  const resp = await apiFetch(path);
  const text = await resp.text();
  if (!resp.ok) {
    throw new ApiError(resp.status, extractErrorMessage(resp.status, text, resp.statusText));
  }
  return text;
}

async function fetchBlob(path: string): Promise<Blob> {
  const resp = await apiFetch(path);
  if (!resp.ok) {
    let text: string;
    try {
      text = await resp.text();
    } catch {
      text = resp.statusText;
    }
    throw new ApiError(resp.status, extractErrorMessage(resp.status, text, resp.statusText));
  }
  return resp.blob();
}

// JSON endpoints

export async function getTreeSize(caId?: string): Promise<MtcTreeSize> {
  return apiJson(`/admin/mtc/tree-size${caQuery(caId)}`);
}

export async function getRoot(caId?: string): Promise<MtcRoot> {
  return apiJson(`/admin/mtc/root${caQuery(caId)}`);
}

export async function getLandmarks(caId?: string): Promise<MtcLandmark[]> {
  const resp: { landmarks: MtcLandmark[] } = await apiJson(`/admin/mtc/landmarks${caQuery(caId)}`);
  return resp.landmarks;
}

export async function getInclusionProof(certId: string): Promise<MtcInclusionProof> {
  return apiJson(`/admin/mtc/inclusion-proof/${encodeURIComponent(certId)}`);
}

export async function getConsistencyProof(
  caId: string,
  from: number,
  to: number,
): Promise<MtcConsistencyProof> {
  const params = new URLSearchParams({
    ca_id: caId,
    from: String(from),
    to: String(to),
  });
  return apiJson(`/admin/mtc/consistency-proof?${params}`);
}

export async function getSubtreeRoot(
  caId: string,
  start: number,
  end: number,
): Promise<MtcSubtreeRoot> {
  const params = new URLSearchParams({
    ca_id: caId,
    start: String(start),
    end: String(end),
  });
  return apiJson(`/admin/mtc/subtree-root?${params}`);
}

export async function getRevokedRanges(caId?: string): Promise<MtcRevokedRange[]> {
  const resp: { revoked_ranges: MtcRevokedRange[] } = await apiJson(`/admin/mtc/revoked-ranges${caQuery(caId)}`);
  return resp.revoked_ranges;
}

// Text endpoints

export async function getLandmarkList(caId?: string): Promise<string> {
  return fetchText(`/admin/mtc/landmark-list${caQuery(caId)}`);
}

export async function getCheckpoint(caId?: string): Promise<string> {
  return fetchText(`/admin/mtc/checkpoint${caQuery(caId)}`);
}

export async function getCosignature(caId?: string): Promise<string> {
  return fetchText(`/admin/mtc/cosignature${caQuery(caId)}`);
}

// Binary downloads

export async function downloadStandalone(certId: string): Promise<Blob> {
  return fetchBlob(`/admin/mtc/standalone/${encodeURIComponent(certId)}`);
}

export async function downloadLandmarkCert(seq: number, caId?: string): Promise<Blob> {
  return fetchBlob(`/admin/mtc/landmarks/${seq}/cert${caQuery(caId)}`);
}

export interface MtcLandmarkCertDetails {
  sequence_no: number;
  cert_text: string | null;
}

export async function getLandmarkCertDetails(
  seq: number,
  caId?: string,
): Promise<MtcLandmarkCertDetails> {
  return apiJson(`/admin/mtc/landmarks/${seq}/cert-details${caQuery(caId)}`);
}

export async function getLogListEntry(caId: string): Promise<string> {
  return fetchText(`/admin/ca/${encodeURIComponent(caId)}/mtc/log-list-entry`);
}

// Actions

export async function forceCheckpoint(caId: string): Promise<void> {
  return apiVoid(`/admin/ca/${encodeURIComponent(caId)}/mtc/force-checkpoint`, { method: 'POST' });
}

export async function forceLandmark(caId: string): Promise<void> {
  return apiVoid(`/admin/ca/${encodeURIComponent(caId)}/mtc/force-landmark`, { method: 'POST' });
}

File: crates/akamuctl/src/commands/mtc.rs
//! MTC transparency log administration subcommands (akamuctl mtc …).

use std::path::PathBuf;

use crate::client::AdminClient;
use crate::error::CtlError;
use crate::output::{print, Format};
use crate::urlenc;

fn ca_qs(ca: &Option<String>) -> String {
    match ca {
        Some(id) => format!("?ca_id={}", urlenc(id)),
        None => String::new(),
    }
}

/// Show MTC transparency log tree size.
pub async fn tree_size(
    client: &AdminClient,
    fmt: &Format,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let resp = client
        .get(&format!("/admin/mtc/tree-size{}", ca_qs(&ca)))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// Show MTC tree size and root hash.
pub async fn root(client: &AdminClient, fmt: &Format, ca: Option<String>) -> Result<(), CtlError> {
    let resp = client
        .get(&format!("/admin/mtc/root{}", ca_qs(&ca)))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// List MTC landmarks as JSON.
pub async fn landmarks(
    client: &AdminClient,
    fmt: &Format,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let resp = client
        .get(&format!("/admin/mtc/landmarks{}", ca_qs(&ca)))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// Show landmarks in spec section 3.4 text format.
pub async fn landmark_list(client: &AdminClient, ca: Option<String>) -> Result<(), CtlError> {
    let text = client
        .get_text(&format!("/admin/mtc/landmark-list{}", ca_qs(&ca)))
        .await?;
    print!("{text}");
    Ok(())
}

/// Download landmark certificate DER by sequence number.
pub async fn landmark_cert(
    client: &AdminClient,
    seq: i64,
    ca: Option<String>,
    output: Option<PathBuf>,
) -> Result<(), CtlError> {
    let bytes = client
        .get_bytes(&format!("/admin/mtc/landmarks/{seq}/cert{}", ca_qs(&ca)))
        .await?;
    write_binary(&bytes, output.as_deref())
}

/// Show parsed details of a landmark certificate.
pub async fn landmark_cert_show(
    client: &AdminClient,
    seq: i64,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let bytes = client
        .get_bytes(&format!("/admin/mtc/landmarks/{seq}/cert{}", ca_qs(&ca)))
        .await?;
    match akamu_client::cert_text::describe_landmark_cert_der(&bytes) {
        Some(text) => print!("{text}"),
        None => eprintln!("Failed to parse landmark certificate DER"),
    }
    Ok(())
}

/// Show inclusion proof for a certificate.
pub async fn inclusion_proof(
    client: &AdminClient,
    fmt: &Format,
    cert_id: &str,
) -> Result<(), CtlError> {
    let resp = client
        .get(&format!("/admin/mtc/inclusion-proof/{}", urlenc(cert_id)))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// Download standalone DER certificate.
pub async fn standalone(
    client: &AdminClient,
    cert_id: &str,
    output: Option<PathBuf>,
) -> Result<(), CtlError> {
    let bytes = client
        .get_bytes(&format!("/admin/mtc/standalone/{}", urlenc(cert_id)))
        .await?;
    write_binary(&bytes, output.as_deref())
}

/// Show consistency proof between two tree sizes.
pub async fn consistency_proof(
    client: &AdminClient,
    fmt: &Format,
    from: u64,
    to: u64,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let mut qs = format!("?from={from}&to={to}");
    if let Some(ref id) = ca {
        qs.push_str(&format!("&ca_id={}", urlenc(id)));
    }
    let resp = client
        .get(&format!("/admin/mtc/consistency-proof{qs}"))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// Compute subtree root hash over a leaf range.
pub async fn subtree_root(
    client: &AdminClient,
    fmt: &Format,
    start: u64,
    end: u64,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let mut qs = format!("?start={start}&end={end}");
    if let Some(ref id) = ca {
        qs.push_str(&format!("&ca_id={}", urlenc(id)));
    }
    let resp = client.get(&format!("/admin/mtc/subtree-root{qs}")).await?;
    print(fmt, &resp);
    Ok(())
}

/// Show revoked leaf-index ranges.
pub async fn revoked_ranges(
    client: &AdminClient,
    fmt: &Format,
    ca: Option<String>,
) -> Result<(), CtlError> {
    let resp = client
        .get(&format!("/admin/mtc/revoked-ranges{}", ca_qs(&ca)))
        .await?;
    print(fmt, &resp);
    Ok(())
}

/// Show C2SP tlog operator checkpoint.
pub async fn checkpoint(client: &AdminClient, ca: Option<String>) -> Result<(), CtlError> {
    let text = client
        .get_text(&format!("/admin/mtc/checkpoint{}", ca_qs(&ca)))
        .await?;
    print!("{text}");
    Ok(())
}

/// Show C2SP tlog cosignature checkpoint.
pub async fn cosignature(client: &AdminClient, ca: Option<String>) -> Result<(), CtlError> {
    let text = client
        .get_text(&format!("/admin/mtc/cosignature{}", ca_qs(&ca)))
        .await?;
    print!("{text}");
    Ok(())
}

/// Force an immediate MTC checkpoint for the given CA.
pub async fn force_checkpoint(client: &AdminClient, ca: &str) -> Result<(), CtlError> {
    client
        .post_action(&format!("/admin/ca/{}/mtc/force-checkpoint", urlenc(ca)))
        .await?;
    println!("Checkpoint produced for CA '{ca}'.");
    Ok(())
}

/// Force an immediate MTC landmark allocation for the given CA.
pub async fn force_landmark(client: &AdminClient, ca: &str) -> Result<(), CtlError> {
    client
        .post_action(&format!("/admin/ca/{}/mtc/force-landmark", urlenc(ca)))
        .await?;
    println!("Landmark allocated for CA '{ca}'.");
    Ok(())
}

fn write_binary(bytes: &[u8], output: Option<&std::path::Path>) -> Result<(), CtlError> {
    if let Some(path) = output {
        std::fs::write(path, bytes)?;
        println!("Written {} bytes to {}", bytes.len(), path.display());
    } else {
        for (i, b) in bytes.iter().enumerate() {
            if i > 0 && i % 32 == 0 {
                println!();
            }
            print!("{b:02x}");
        }
        println!();
    }
    Ok(())
}

/// Print the Witness Network log-list entry for this CA's MTC log.
pub async fn log_list_entry(client: &AdminClient, ca: &str) -> Result<(), CtlError> {
    let text = client
        .get_text(&format!("/admin/ca/{}/mtc/log-list-entry", urlenc(ca)))
        .await?;
    print!("{text}");
    Ok(())
}
